app/com/arpnetworking/rollups/QueryConsistencyTaskCreator.java

```diff
@@ -0,0 +1,172 @@
+ * You may obtain a copy of the License at
+ *
+ *     http://www.apache.org/licenses/LICENSE-2.0
+ *
+ * Unless required by applicable law or agreed to in writing, software
+ * distributed under the License is distributed on an "AS IS" BASIS,
+ * WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the License for the specific language governing permissions and
+ * limitations under the License.
+ */
+package com.arpnetworking.rollups;
+
+import akka.actor.ActorRef;
+import akka.pattern.Patterns;
+import com.arpnetworking.kairos.client.models.Metric;
+import com.arpnetworking.kairos.client.models.MetricsQuery;
+import com.arpnetworking.steno.Logger;
+import com.arpnetworking.steno.LoggerFactory;
+
+import java.time.Duration;
+import java.time.Instant;
+import java.util.Iterator;
+import java.util.Random;
+import java.util.Spliterator;
+import java.util.Spliterators;
+import java.util.concurrent.ExecutionException;
+import java.util.function.Consumer;
+import java.util.stream.Stream;
+import java.util.stream.StreamSupport;
+
+/**
+ * Randomly samples some queries to be forwarded to the rollup consistency checker (if they contain rollups at all).
+ *
+ * @author William Ehlhardt (whale at dropbox dot com)
+ */
+public class QueryConsistencyTaskCreator implements Consumer<MetricsQuery> {
+    private static final Logger LOGGER = LoggerFactory.getLogger(QueryConsistencyTaskCreator.class);
+
+    private static final Random RANDOM = new Random();
+
+    private final double _checkFraction;
+    private final ActorRef _consistencyChecker;
+
+    /**
+     * Constructor.
+     *
+     * @param checkFraction Fraction of queries to send for checking.
+     * @param consistencyChecker Destination consistency checker actor.
+     */
+    public QueryConsistencyTaskCreator(final double checkFraction, final ActorRef consistencyChecker) {
+        this._checkFraction = checkFraction;
+        this._consistencyChecker = consistencyChecker;
+    }
+
+    @Override
+    public void accept(final MetricsQuery query) {
+        if (RANDOM.nextDouble() > _checkFraction) {
+            return;
+        }
+
+        if (!query.getStartTime().isPresent()) {
+            LOGGER.trace()
+                    .setMessage("not consistency-checking because no start time present")
+                    .addData("query", query)
+                    .log();
+            return;
+        }
+
+        if (!query.getEndTime().isPresent()) {
+            LOGGER.trace()
+                    .setMessage("not consistency-checking because no end time present")
+                    .addData("query", query)
+                    .log();
+            return;
+        }
+
+        LOGGER.trace()
+                .setMessage("maybe sending for consistency check?")
+                .addData("query", query)
+                .log();
+
+        final Instant startTime = query.getStartTime().get();
+        final Instant endTime = query.getEndTime().get();
```

Comment:
I think we want to match KairosDB's semantics here:
```suggestion
        final Instant endTime = query.getEndTime().orElse(Instant.now());
```
(and ditch the `if (!query.getEndTime().isPresent()) {...}` block above)